nodejs/aws-infra/experimental/image.ts

```diff
@@ -0,0 +1,206 @@
+// Copyright 2016-2018, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import * as aws from "@pulumi/aws";
+import * as docker from "@pulumi/docker";
+import * as pulumi from "@pulumi/pulumi";
+
+import * as utils from "../utils";
+
+import * as mod from ".";
+
+export type ImageEnvironment = Record<string, pulumi.Input<string>>;
+
+export interface IImageProvider {
+    image(name: string, parent: pulumi.Resource): pulumi.Input<string>;
+    environment(name: string, parent: pulumi.Resource): pulumi.Input<aws.ecs.KeyValuePair[]>;
+}
```

Comment:
a 'provider' is a general purpose way of flipping the onus of data production around.  instead of something like service/task needing to know how to compute this, we simply hide it behind this interface.  

Technically, i could have implemented this witout even passing the ImageProvider to something like TaskDefinition.  *However*, in that case, the ImageProvider would ahve to make the resources without having a clear idea about who should be the parent.  Instead, we allow the parent to be passed in here, allowing the TaskDefinition to pass itself in here.  This gives good parenting.

If we decide this is not important, we can simply make image/environment into properties.

Reply:
Note: providers should have methods that return 'Inputs'.  this gives impls the most flexibility in terms of computing and returning data.